components/core/src/clp/ffi/SchemaTree.hpp

```diff
@@ -0,0 +1,151 @@
+#ifndef CLP_FFI_SCHEMATREE_HPP
+#define CLP_FFI_SCHEMATREE_HPP
+
+#include <cstddef>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "../ErrorCode.hpp"
+#include "../TraceableException.hpp"
+#include "SchemaTreeNode.hpp"
+
+namespace clp::ffi {
+/**
+ * This class implements a schema tree, providing all necessary methods for upper-layer APIs through
+ * FFI. Nodes are stored using a flattened vector.
+ */
+class SchemaTree {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(
+                ErrorCode error_code,
+                char const* const filename,
+                int line_number,
+                std::string message
+        )
+                : TraceableException{error_code, filename, line_number},
+                  m_message{std::move(message)} {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return m_message.c_str();
+        }
+
+    private:
+        std::string m_message;
+    };
+
+    /**
+     * When constructing the schema tree, we uniquely identify the location of a node being inserted
+     * to the tree by the unique triple of the parent id, the key name, and the node type. The
+     * reason why the triple is unique is because the combination of the key name and the node type
+     * should have no ambiguity for a parent node. This class stores such a triple and act as a
+     * unique identifier for a node in the schema tree.
```

Comment:
If we adopt the new class docstring, then I think we can simplify this:

```suggestion
     * A triple---parent ID, key name, and node type---that uniquely identifies a node.
     * NOTE: We use the term "Locator" to avoid terms like "Key" or "Identifier" that are already in
     * use.
```